### ISSUE TITLE
FilterChain Discovery Service (FDS)

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#4540]
In order to support filter chain updates without forcing a full listener drain, it's been suggested we add a FilterChain Discovery Service. This would allow individual FilterChains belonging to a Listener to be independently discovered and updated.

This would also lend support to finer grained Listener resource naming for incremental and on-demand xDS (generalizing https://github.com/envoyproxy/envoy/issues/2500).

Opening this issue to track any design work and ownership.

@ggreenway @andraxylia @lizan @PiotrSikora @mattklein123

**Comment by andraxylia** [comment]
We need this fix for Istio for long-lived connections with SNI @rshriram @costinm .

**Comment by ggreenway** [comment]
Related add-on feature: lazy-loading of FilterChains, triggered by a listener-filter that is watching for unknown SNI values.

**Comment by andraxylia** [comment]
@htuch does the drain happen when a particular FilterChain is updated, and also when any FilterChain is added or deleted in the listener?

**Comment by mattklein123** [comment]
> We need this fix for Istio for long-lived connections with SNI

@andraxylia the intention here is to only effect new connections, right? Obviously previous connections will use the old filter chain? I just want to make sure we are on the same page. If so this sounds like a useful and straightforward feature to me.

**Comment by andraxylia** [comment]
@mattklein123 All is good if a change in the FilterChain  does not affect existing TCP and GRPC connections - for instance the addition of a host in server_names in the FilterMatch.

@rshriram @costinm told me that because of issues with full drain we cannot move Istio from multiple listeners to 1 listener with multiple FilterChains.

**Comment by ggreenway** [comment]
I brainstormed with @htuch and here's what we propose for now:

* Listeners will specify a list of FilterChains by name to discover. That means that adding a new static FilterChain will involve modifying the listener. This model is the best way to support lazy-loading of FilterChains.
* We will modify the current listener implementation so that if a listener has filter chains added, modified, or removed, draining only happens to connections on the affected filter chain. Other connections do not need to drain. This will be done as a separate PR, as it does not require filterchain discovery.
* FilterChain Discovery Service is geared towards the lazy-load use case. Statically defined/loaded filter chains can be added/modified/removed with the existing LDS.

Any comments or concerns with this plan?

**Comment by mattklein123** [comment]
I don't understand how filter chains interact with draining? Can't old connections use the old filter chain and new ones use the new filter chain? If we want to add draining on filter chain change I think that's totally orthogonal?

**Comment by ggreenway** [comment]
@mattklein123 Yes, that's basically the conclusion @htuch and I came to. There are two orthogonal feature requests/requirements here:

1) Be able to add/remove filterchains without draining. Right now this is not possible only due to the implementation of listener config changes. The API does not need to change to fix this.
2) lazy-load filterchains.

The implementation of these two features is mostly/entirely non-intersecting.

**Comment by mattklein123** [comment]
OK, thanks for the clarification. IMO we don't ever need to implement draining on a per filter chain basis. I would track as a totally separate issue? Fixing (1) should be simple compared to having to do draining also (not simple).

**Comment by mattklein123** [comment]
But even with lazy load, why is draining required? It seems like we can write the code so that a filter chain is latched onto a connection at initial connect time? Why is it required to drain? (I can see why someone might want that, but it seems not required).

**Comment by ggreenway** [comment]
That's an interesting point. I suppose it's not required. But it would be inconsistent with current behavior for listeners: there's not currently a way to modify/remove a listener without draining. But the answer to (1) may be as simple as adding a new DrainType of "hot-restart only". That's not quite as fine-grained, but would be much less effort for most of the potential gains. It would entirely solve the case of filterchain add/remove, but would not allow forcing draining for a filterchain modify.

**Comment by mattklein123** [comment]
FWIW, my view of filter chain is that it's something that is only consulted when a connection is created. As such, draining is not required. The reason I wrote draining into listeners is that when a listener goes away, by definition is destroys all owned connections (this wouldn't be the case for filter chains). I agree that we could add a configuration option to listeners to not drain and just hard swap, though it seems like few would want this behavior?

**Comment by mattklein123** [comment]
> I don't think I understand what you have in mind for the different lifetimes of listeners vs filter chains and why they're fundamentally different.

The reason that I added draining for the listener implementation was honestly mainly for simplicity due to connection ownership. I agree that connections could live beyond the life of a listener but it would be a very large change. Obviously not doing anything at all (draining vs. shared_ptr ownership) would be not good for many users.

For filter chains, I think the keeping the filter chain alive via shared pointer if necessary is substantially simpler and if it were me is how I would start. I don't object to adding draining in any way, I just don't think it adds much and would be pretty complicated.

**Comment by ggreenway** [comment]
Ok, that makes sense. I just wanted to make sure I understood your reasoning.

What do you think about changing listeners so that they do not drain when their config is modified (but not deleted) if the only part of the config that changed is the filter chains?

**Comment by mattklein123** [comment]
> What do you think about changing listeners so that they do not drain when their config is modified (but not deleted) if the only part of the config that changed is the filter chains?

Yup makes sense and I don't think would be that hard.

**Comment by htuch** [comment]
SGTM to me as well, FWIW I hadn't grokked the full tradeoff between shared_ptr and drain here in Listener implementation, so thanks @mattklein123 for explaining this for posterity.

**Comment by ggreenway** [comment]
@mattklein123 I've been thinking this through, and simply using a shared_ptr for the filter-chain will run into other issues, because right now ThreadLocal slots must be destructed on the main thread. But I can probably make it post deletions to the main thread when the refcount goes to zero somehow.

**Comment by mattklein123** [comment]
Hmm yeah that's definitely a complexity with the stored lambdas. Yeah I think could have some type of shared_ptr wrapper that on destruction posts back to the main listener for deletion. It's definitely not trivial b but doable. I suppose if TLS is the only issue you could fix that to allow deletion from any thread, but not sure if that is worth it or not.

**Comment by rshriram** [comment]
my two cents. the draining is totally optional. If we can have the old filter chain with the old connection live for ever in the same listener, then as Matt says, I see no reason to drain. Unless its a security issue and I want to terminate old connections in a timely fashion, and not have any connections with expired tls certs lingering around for ages.

**Comment by PiotrSikora** [comment]
Wouldn't not draining lead to a situation, where we have long-running connections (i.e. HTTP/2, gRPC) that are latched to a filter chain that no longer exist, along with its outdated configuration, with no way to force those clients to re-connect and re-select a new filter chain?

**Comment by rshriram** [comment]
it does but the point is that unless these configurations pertain to some security stuff, this "forcing" thingy should be optional. In a way, we put the onus on the end user - if they want to take advantage of the newer config, they should reconnect. Gives them the option of staying with old config or reconnecting

**Comment by mattklein123** [comment]
@PiotrSikora I don't think anyone is saying we shouldn't implement filter chain draining. All we are saying is that it's extra complexity that is orthogonal to the main implementation. I would suggest we track as a separate issue once the primary feature is implemented.

**Comment by silentdai** [comment]
Ack: working on this

**Comment by davecheney** [comment]
@silentdai would you be able to post a brief status update on this issue for followers. Thanks. Dave

**Comment by silentdai** [comment]
#7152 #7246 are the preparation. Not a real fdcs thing but share the motivation that decouple the filterchain and listener flow.
I have several draft PRs on benchmark and FCDS proto. Blocked by #7246 
I also have draft design doc supposed to publish by the end of this week.

**Comment by PiotrSikora** [comment]
One more thing to consider as part of this work is decoupling of filter chains from the filter chain matchers. This would not only help with the future on-demand work, but it would also allow multiple filter chain matchers to match to a single filter chain, instead of having separate instances of filter chains for each matcher, since those can be exact copies in some cases.

**Comment by rgs1** [comment]
fwiw, for our use case it would be great to allow multiple matchers to refer to the same chain and avoid the extra copies.

**Comment by htuch** [comment]
@mattklein123 @lambdai do you think this is needed for 1.12.0 or can we do this as a v3 add-on once shipped?

**Comment by mattklein123** [comment]
Add on is fine with me.

**Comment by hirenshah05** [comment]
I was curious about time-line on this feature (FDS). Looks like lot of work is already committed.

**Comment by jonmeacham** [comment]
@lambdai thank you for your work on this! It looks like the result may have landed in [1.15.0](https://www.envoyproxy.io/docs/envoy/v1.15.0/version_history/current). Am I mistaken?

`listener: added in place filter chain update flow for tcp listener update which doesn’t close connections if the corresponding network filter chain is equivalent during the listener update. Can be disabled by setting runtime feature envoy.reloadable_features.listener_in_place_filterchain_update to false. Also added additional draining filter chain stat for listener manager to track the number of draining filter chains and the number of in place update attempts.`

**Comment by vorishirne** [comment]
@htuch will this support on-demand filter-chain, based on SNI or similar param

**Comment by htuch** [comment]
We've pivoted towards not doing filter-chain discovery, instead relying on improvements to LDS update (i.e. drainless updates) and ECDS. I'm going to close this out, as I think the original goals we were shooting towards are satisfied by the aforementioned. Feel free to reopen if this is not the case.

**Comment by phylake** [comment]
Perhaps a goal that I should have mentioned sooner and the reason I've been tracking this issue is to reduce the blast radius of a LDS NACK. My current NACK handling breaks up Resources into individual DeltaDiscoveryResponses to see which of the Resources produced the NACK and not resend it. This works well for granular resources like SDS, EDS, and CDS which may only affect a single service.

A LDS NACK affects every service with traffic flowing through its port. For me that's thousands of services on a single compute cluster

**Comment by htuch** [comment]
@phylake what typically causes these NACKs? I'm wondering if you move to ECDS whether enough of the config is out of the `Listener` resource that this will make a material difference. If not, maybe we should reopen.

**Comment by vorishirne** [comment]
> I think the original goals we were shooting towards are satisfied by the aforementioned. Feel free to reopen if this is not the case.

We will never have on-demand update of filter chain.

**Comment by htuch** [comment]
Have you considered doing more extensive validation offline or via canary? I think what you describe is a valid reason to have finer grained resources in the model, but other mechanisms to ensure that NACK is unlikely to happen will pay off more broadly.

**Comment by htuch** [comment]
I'm going to reopen as this is a valid consideration, plus if we did want to lazy load the filter chains, we would need this.

**Comment by vorishirne** [comment]
again, can the on-demand loading of filter chain be a part of this? If everything is on-demand, can u please make this also.

**Comment by lambdai** [comment]
I'd like to revisit this after [unified filter chain matcher](https://github.com/envoyproxy/envoy/pull/20110) is adopted in istio.

That API enforces the filter chain name. This change greatly reduced the complexity of FDS

**Comment by ssripadham** [comment]
I see a few benefits of developing FCDS.

1. It Allows us modeling customizations built into a single listener.
    Example:
     Listener
           FilterChain1: Some unique rules in fcd1.conf
           FilterChain2: Some unique rules in fcds2.conf
          ...
3. Support breaking a single LDS file into more manageable multiple FCDS files. Imagine these FCDS files holding several lines (1K+) of envoy yaml.
4. Filter Chain discovery & FilterChain Only updates (This feature exists as I understand in the lds itself) 
5. Filter chains can be loaded/unloaded on demand. This way all of the Filter Chains need not be held in process memory.
6. Any other benefit?

Is this feature actively worked on?

**Comment by rakeshdatta** [comment]
Folks, this is the implementation PR for FCDS (SoTW). Plz review. 
Pending work: 

1. To tests and possible hook some more changes for FC level connection draining.
2. Delta updates for FCDS
